### PR TITLE
feat: 🍰 Contribution Messages In Frontend

### LINKED ISSUE #2132: 🚀 [Feature] Messages in Wallet
## 🚀 Feature
Allow a user to add messages to own contributions, that are not deleted and not confirmed and not denied.

Clarify the conditions a user can delete a message.

Clarify if messages can be edited.

Indicate contributions that have messages.

List all messages after clicking on the contribution.

### LINKED ISSUE #2133: 🚀 [Feature] Messages in Admin Interface
## 🚀 Feature
Allow admins to add messages to contributions.

List messages to given contribution.

### PR DESCRIPTION
**Pull request by ogerly**
## 🍰 Pullrequest
The Contribution Messages System for frontend and admin area. 

### graphql
- creat contribution message for user
- creat contribution message for admin

### frontend
- messages in contribution list

### Issues
 - fixes #2132 
 - fixes #2133 

 

### Merged
- merge #2149 
- merge #2161  
- merge #2142

### CODE REVIEW

**Comment by ogerly** [comment]
FireShot Capture 1244 - Gradido Admin Interface - localhost

FireShot Capture 1243 - Gradido Admin Interface - localhost

FireShot Capture 1242 -  - localhost

FireShot Capture 1241 -  - localhost

**Comment by ogerly** [comment]
https://user-images.githubusercontent.com/1324583/187176838-056f4aa7-7f4e-434e-b499-59dc1f03bf09.mp4

File: admin/src/locales/en.json [hunk 1]
```diff
@@ -69,6 +69,10 @@
     },
     "short_hash": "({shortHash})"
   },
+  "form": {
+    "reset": "reset",
+    "submit": "submit"
```
Comment (Tirokk) [inline admin/src/locales/en.json:74]:
```suggestion
    "reset": "Cancel",
    "submit": "Send"
```

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.vue [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        <b-form-textarea
+          id="textarea"
+          v-model="form.text"
+          placeholder="Enter something..."
+          rows="3"
+          max-rows="6"
+        ></b-form-textarea>
+        <b-row class="mt-4 mb-6">
+          <b-col>
+            <b-button type="reset" variant="danger">{{ $t('form.reset') }}</b-button>
```
Comment (Tirokk) [inline frontend/src/components/ContributionMessages/ContributionMessagesFormular.vue:14]:
```suggestion
            <b-button type="reset" variant="danger">{{ $t('form.cancel') }}</b-button>
```

Like on `ContributionForm`?

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            form: {
+              text: 'text form message',
+            },
+          })
+          wrapper.vm.onReset()
+        })
+        expect(wrapper.vm.form).toEqual({
+          text: '',
+        })
+      })
```
Comment (Tirokk) [inline frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:56]:
```suggestion
      beforeEach(() => {
        wrapper.setData({
          form: {
            text: 'text form message',
          },
        })
        wrapper.vm.onReset()
      })

      it('form has the set data', () => {
        expect(wrapper.vm.form).toEqual({
          text: '',
        })
      })
```

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  describe('mount', () => {
+    beforeEach(() => {
+      wrapper = Wrapper()
+    })
+
+    it('has a DIV .contribution-messages-formular', () => {
+      expect(wrapper.find('div.contribution-messages-formular').exists()).toBe(true)
+    })
+
+    describe('call onReset', () => {
```
Comment (Tirokk) [inline frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:43]: I would more prefer to `on trigger reset` here like:
https://github.com/gradido/gradido/blob/0da776b2c6eae30f4e790cbc04884b365b6969b7/frontend/src/components/Contributions/ContributionForm.spec.js#L291-310

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+          })
+          wrapper.vm.onReset()
+        })
+        expect(wrapper.vm.form).toEqual({
+          text: '',
+        })
+      })
+    })
+
+    describe('call onSubmit', () => {
```
Comment (Tirokk) [inline frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:59]: I would more prefer to `on trigger submit` here like:
https://github.com/gradido/gradido/blob/0da776b2c6eae30f4e790cbc04884b365b6969b7/frontend/src/components/Contributions/ContributionForm.spec.js#L291-310

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.vue [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  },
+  data() {
+    return {
+      form: {
+        text: '',
+      },
+    }
+  },
+  methods: {
+    onSubmit(event) {
```
Comment (Tirokk) [inline admin/src/components/ContributionMessages/ContributionMessagesFormular.vue:43]: I get an error toasted localy:

Bildschirmfoto 2022-08-30 um 11 09 17

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.vue [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          variables: {
+            contributionId: this.contributionId,
+            message: this.form.text,
+          },
+        })
+        .then((result) => {
+          this.$emit('get-list-contribution-messages', this.contributionId)
+          this.$emit('update-state', this.contributionId)
+          this.form.text = ''
+          this.toastSuccess(this.$t('message.request'))
```
Comment (Tirokk) [inline admin/src/components/ContributionMessages/ContributionMessagesFormular.vue:56]: I get an undefined toast message:

Bildschirmfoto 2022-08-30 um 11 26 53

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.vue [hunk 8]
```diff
@@ -0,0 +1,67 @@
+<template>
+  <div class="contribution-messages-formular">
+    <div>
+      <b-form @submit.prevent="onSubmit" @reset.prevent="onReset">
+        <b-form-textarea
+          id="textarea"
+          v-model="form.text"
+          placeholder="Enter something..."
```
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesFormular.vue:8]: ok

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.spec.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+          text: '',
+        })
+      })
+    })
+
+    describe('call onSubmit', () => {
+      it('response with the contribution message', () => {
+        wrapper.vm.onSubmit()
+      })
+    })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:64]: ok

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      it('response with the contribution message', () => {
+        wrapper.vm.onSubmit()
+      })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:63]: ok

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js (hunk 4 above)
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:43]: ok

File: frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js (hunk 5 above)
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesFormular.spec.js:60]: ok

File: admin/src/components/ContributionMessages/ContributionMessagesList.vue [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    },
+  },
+  data() {
+    return {
+      messages: [],
+    }
+  },
+  methods: {
+    getListContributionMessages(id) {
+      this.messages = []
```
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesList.vue:41]: I forgot to delete this

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.vue (hunk 7 above)
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesFormular.vue:56]: is changed

File: admin/src/components/ContributionMessages/ContributionMessagesFormular.vue (hunk 6 above)
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesFormular.vue:43]: is changed

File: frontend/src/components/Contributions/ContributionListItem.vue [hunk 12]
```diff
@@ -2,47 +2,100 @@
...
-            "
+            v-if="(state !== 'DELETED' && !firstName) || (state !== 'CONFIRMED' && !firstName)"
+            class="d-flex flex-row-reverse"
           >
-            <b-icon icon="pencil" class="h2"></b-icon>
-          </div>
-          <div class="pointer" @click="deleteContribution({ id })">
-            <b-icon icon="trash" class="h2"></b-icon>
+            <div
+              v-if="state !== 'CONFIRMED' && state !== 'DELETED'"
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionListItem.vue:35]:
```suggestion
              v-if="!['CONFIRMED', 'DELETED'].includes(state)"
```

See below … 👇🏼

File: frontend/src/components/Contributions/ContributionListItem.vue [hunk 13]
```diff
@@ -2,47 +2,100 @@
...
-            class="pointer ml-5"
-            @click="
-              $emit('update-contribution-form', {
-                id: id,
-                contributionDate: contributionDate,
-                memo: memo,
-                amount: amount,
-              })
-            "
+            v-if="(state !== 'DELETED' && !firstName) || (state !== 'CONFIRMED' && !firstName)"
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionListItem.vue:31]:
```suggestion
            v-if="!['CONFIRMED', 'DELETED'].includes(state) && !firstName"
```

Alternative code. Hope this is correct …

I would not decide by the existenz of the "firstName" to show the edit, trash or message buttons!
Please make a property for the `ContributionList` and from there to the `ContributionListItem` which decides if the buttons should be shown …

File: frontend/src/components/Contributions/ContributionListItem.vue [hunk 14]
```diff
@@ -2,47 +2,100 @@
...
+            {{ $t('contribution.alert.answerQuestion') }}
+          </b-button>
+          <b-collapse :id="collapsId" class="mt-2">
+            <b-card>
+              <contribution-messages-list
+                :messages="messages_get"
+                :state="state"
+                :contributionId="contributionId"
+                @get-list-contribution-messages="getListContributionMessages"
+                @update-state="updateState"
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionListItem.vue:81]: Please test for reemitation of this event like in all the other examples I gave you above.

File: frontend/src/locales/en.json [hunk 15]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Activity",
     "alert": {
+      "answerQuestion": "Please answer the question",
       "communityNoteList": "Here you will find all submitted and confirmed contributions from all members of this community.",
       "confirm": "confirmed",
+      "in_progress": "Es gibt eine Rückfrage der Moderatoren.",
```
Comment (Tirokk) [inline frontend/src/locales/en.json:41]:
```suggestion
      "in_progress": "There is a question from the moderators.",
```

File: frontend/src/locales/es.json [hunk 16]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Actividad",
     "alert": {
+      "answerQuestion": "Please answer the question",
       "communityNoteList": "Aquí encontrarás todas las contribuciones enviadas y confirmadas de todos los miembros de esta comunidad.",
       "confirm": "confirmado",
+      "in_progress": "Es gibt eine Rückfrage der Moderatoren.",
```
Comment (Tirokk) [inline frontend/src/locales/es.json:41]:
```suggestion
      "in_progress": "There is a question from the moderators.",
```

May @Mogge could translate it?

File: frontend/src/locales/es.json [hunk 17]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Actividad",
     "alert": {
+      "answerQuestion": "Please answer the question",
```
Comment (Tirokk) [inline frontend/src/locales/es.json:38]: May @Mogge could translate it?

File: frontend/src/locales/fr.json [hunk 18]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Activité",
     "alert": {
+      "answerQuestion": "Please answer the question",
       "communityNoteList": "Vous trouverez ci-contre toutes les contributions versées et certifiées de tous les membres de cette communauté.",
       "confirm": " Approuvé",
+      "in_progress": "Es gibt eine Rückfrage der Moderatoren.",
```
Comment (Tirokk) [inline frontend/src/locales/fr.json:41]:
```suggestion
      "in_progress": "There is a question from the moderators.",
```

May @Elweyn  could translate it?

File: frontend/src/locales/fr.json [hunk 19]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Activité",
     "alert": {
+      "answerQuestion": "Please answer the question",
```
Comment (Tirokk) [inline frontend/src/locales/fr.json:38]: May @Elweyn could translate it?

File: frontend/src/locales/nl.json [hunk 20]
```diff
@@ -34,10 +35,11 @@
   "contribution": {
     "activity": "Activiteit",
     "alert": {
+      "answerQuestion": "Please answer the question",
       "communityNoteList": "Hier vind je alle ingediende en bevestigde bijdragen van alle leden uit deze gemeenschap.",
       "confirm": "bevestigt",
+      "in_progress": "Es gibt eine Rückfrage der Moderatoren.",
```
Comment (Tirokk) [inline frontend/src/locales/nl.json:41]:
```suggestion
      "in_progress": "There is a question from the moderators.",
```

File: frontend/src/pages/Community.spec.js [hunk 21]
```diff
@@ -92,10 +92,6 @@ describe('Community', () => {
       it('has three tabs', () => {
         expect(wrapper.findAll('div[role="tabpanel"]')).toHaveLength(3)
       })
-
-      it('has first tab active by default', () => {
-        expect(wrapper.findAll('div[role="tabpanel"]').at(0).classes('active')).toBe(true)
-      })
```
Comment (Tirokk) [inline frontend/src/pages/Community.spec.js:98]: Why this test is removed?

File: frontend/src/pages/Community.vue [hunk 22]
```diff
@@ -2,7 +2,7 @@
   <div class="community-page">
     <div>
       <b-tabs v-model="tabIndex" content-class="mt-3" align="center">
-        <b-tab :title="$t('community.submitContribution')" active>
```
Comment (Tirokk) [inline frontend/src/pages/Community.vue:5]: Why the first tab is not activated by default?

File: frontend/src/pages/Community.vue [hunk 23]
```diff
@@ -258,6 +265,9 @@ export default {
     updateTransactions(pagination) {
       this.$emit('update-transactions', pagination)
     },
+    updateState(id) {
+      this.items.find((item) => item.id === id).state = 'PENDING'
+    },
```
Comment (Tirokk) [inline frontend/src/pages/Community.vue:270]: This update should be tested …

File: frontend/src/pages/Community.vue [hunk 24]
```diff
@@ -271,6 +281,7 @@ export default {
       pageSize: this.pageSize,
     })
     this.updateTransactions(0)
+    this.tabIndex = 1
```
Comment (Tirokk) [inline frontend/src/pages/Community.vue:284]: Is this the activation of the first tab?

**Review by Tirokk — CHANGES_REQUESTED** [review]
Whow !!!
This is a big chunk of work!
Really really cool 😍

So no suprise that I have some changes to suggest and that i have some questions.

Exclamation mark icon it missing:

Bildschirmfoto 2022-08-31 um 12 11 39

File: frontend/src/components/ContributionMessages/slots/IsNotModerator.vue [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+  props: {
+    message: {
+      type: Object,
+      default() {
+        return {}
+      },
+    },
+  },
+  computed: {
+    userName() {
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/ContributionMessages/slots/IsNotModerator.vue]: done

File: frontend/src/components/Contributions/ContributionList.vue [hunk 26]
```diff
@@ -3,8 +3,10 @@
     <div class="list-group" v-for="item in items" :key="item.id">
       <contribution-list-item
         v-bind="item"
+        :contributionId="item.id"
         @update-contribution-form="updateContributionForm"
         @delete-contribution="deleteContribution"
+        @update-state="updateState"
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/Contributions/ContributionList.vue:9]: example:

 describe('delete contribution', () => {
      beforeEach(() => {
        wrapper
          .findComponent({ name: 'ContributionListItem' })
          .vm.$emit('delete-contribution', { id: 2 })
      })

      it('emits delete contribution', () => {
        expect(wrapper.emitted('delete-contribution')).toEqual([[{ id: 2 }]])
      })
    })

File: admin/src/components/ContributionMessages/ContributionMessagesList.spec.js [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+    })
+
+    it('has a DIV .contribution-messages-list', () => {
+      expect(wrapper.find('div.contribution-messages-list').exists()).toBe(true)
+    })
+
+    it('has a Component ContributionMessagesFormular', () => {
+      expect(wrapper.findComponent({ name: 'ContributionMessagesFormular' }).exists()).toBe(true)
+    })
+  })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesList.spec.js:55]: example:

 describe('delete contribution', () => {
      beforeEach(() => {
        wrapper
          .findComponent({ name: 'ContributionListItem' })
          .vm.$emit('delete-contribution', { id: 2 })
      })

      it('emits delete contribution', () => {
        expect(wrapper.emitted('delete-contribution')).toEqual([[{ id: 2 }]])
      })
    })

File: frontend/src/pages/Community.vue (hunk 24 above)
Comment (ogerly) [reply frontend/src/pages/Community.vue:284]: done

File: admin/src/components/ContributionMessages/slots/ContributionMessagesListItem.spec.js [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    })
+
+    it('has a DIV .contribution-messages-list-item', () => {
+      expect(wrapper.find('div.contribution-messages-list-item').exists()).toBe(true)
+    })
+
+    it('props.message.default', () => {
+      expect(wrapper.vm.$options.props.message.default.call()).toEqual({})
+    })
+  })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/slots/ContributionMessagesListItem.spec.js:57]: #2169

File: frontend/src/components/ContributionMessages/ContributionMessagesList.spec.js [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+
+    describe('get List Contribution Messages', () => {
+      beforeEach(() => {
+        wrapper.vm.getListContributionMessages()
+      })
+
+      it('emits getListContributionMessages', async () => {
+        expect(wrapper.vm.$emit('get-list-contribution-messages')).toBeTruthy()
+      })
+    })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesList.spec.js:51]: skiped

File: frontend/src/components/ContributionMessages/ContributionMessagesList.spec.js [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+    describe('update State', () => {
+      beforeEach(() => {
+        wrapper.vm.updateState()
+      })
+
+      it('emits getListContributionMessages', async () => {
+        expect(wrapper.vm.$emit('update-state')).toBeTruthy()
+      })
+    })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesList.spec.js:61]: skiped

File: frontend/src/components/ContributionMessages/ContributionMessagesListItem.spec.js [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe('mount', () => {
+    beforeEach(() => {
+      wrapper = Wrapper()
+    })
+
+    it('has a DIV .contribution-messages-list-item', () => {
+      expect(wrapper.find('div.contribution-messages-list-item').exists()).toBe(true)
+    })
+  })
```
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/ContributionMessages/ContributionMessagesListItem.spec.js:54]: it todo

File: frontend/src/components/Contributions/ContributionList.vue (hunk 26 above)
[the remark this answers is not in the document]
Comment (ogerly) [reply frontend/src/components/Contributions/ContributionList.vue:9]: skiped

File: admin/src/components/ContributionMessages/ContributionMessagesList.spec.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (ogerly) [reply admin/src/components/ContributionMessages/ContributionMessagesList.spec.js:55]: skiped

File: frontend/src/components/Contributions/ContributionListItem.vue (hunk 14 above)
Comment (ogerly) [reply frontend/src/components/Contributions/ContributionListItem.vue:81]: skiped

File: frontend/src/components/ContributionMessages/slots/IsNotModerator.spec.js [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+  describe('mount', () => {
+    beforeEach(() => {
+      wrapper = Wrapper()
+    })
+
+    it('has a DIV .slot-is-not-moderator', () => {
+      expect(wrapper.find('div.slot-is-not-moderator').exists()).toBe(true)
+    })
+
+    it('props.message.default', () => {
```
Comment (Mogge) [inline frontend/src/components/ContributionMessages/slots/IsNotModerator.spec.js:45]: What do you want to test with this? There are more cases

Comment (ogerly) [reply frontend/src/components/ContributionMessages/slots/IsNotModerator.spec.js:45]: in the coverage there was a note at the point where the props object is created as default {}.
 then it wants to have tested it. i tried that. but i only found it somewhere on the net.

File: frontend/src/locales/fr.json (hunk 19 above)
Comment (Elweyn) [reply frontend/src/locales/fr.json:38]:
```suggestion
      "answerQuestion": "S'il te plais répond à la question",
```

File: frontend/src/locales/fr.json [hunk 33]
```diff
@@ -34,10 +35,11 @@
...
       "communityNoteList": "Vous trouverez ci-contre toutes les contributions versées et certifiées de tous les membres de cette communauté.",
       "confirm": " Approuvé",
+      "in_progress": "There is a question from the moderators.",
```
Comment (Elweyn) [inline frontend/src/locales/fr.json:41]: Singular:
```suggestion
      "in_progress": "Il y a une question du modérateur.",
```

**Review by Tirokk — APPROVED** [review]
YEAHHH 🚀🚀🚀💫💫💫

**Pull request merged**